### PR TITLE
Integration of MoonSniff

### PR DESCRIPTION
**Pull request by AP-Frank**
Adds the following:
- MoonSniff scripts for sniffer, postprocessor, traffic generator, and DUT
- Hashmap support
- Writer/Reader for MSCAP files

### CODE REVIEW

File: examples/moonsniff/post-processing.lua [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	-- retrieve file size in order to give an estimation for the processing speed
+	local file = assert(io.open(PRE, "r"))
+        local size = fsize(file)
+        file:close()
+        file = assert(io.open(POST, "r"))
+        size = size + fsize(file)
+        file:close()
+        log:info("File size: " .. size / 1e9 .. " [GB]")
+        local nClock = os.clock()
+        if args.profile then profile.start("-fl", "profile.log") end
```
Comment (emmericp) [inline examples/moonsniff/post-processing.lua]: ws

File: examples/moonsniff/post-processing.lua [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		packets = tbb.match(PRE, POST, args)
+	elseif MODE == MODE_MSCAP then
+		local arr = require "arrmatch"
+		packets = arr.match(PRE, POST, args)
+	end
+
+	-- finish operations
+	if args.profile then profile.stop() end
+
+        local elapsed = os.clock() - nClock
```
Comment (emmericp) [inline examples/moonsniff/post-processing.lua]: avoid using os.clock (it's CPU time, not wall time), use the libmoon timers instead

File: examples/moonsniff/tbbmatch.lua [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+local mod = {}
+
+local memory    = require "memory"
+local ts        = require "timestamping"
+local hist      = require "histogram"
+local log       = require "log"
+local ms	= require "moonsniff-io"
+local dpdk	= require "dpdk"
+local pcap	= require "pcap"
+local hmap	= require "hmap"
```
Comment (emmericp) [inline examples/moonsniff/tbbmatch.lua]: ws

File: examples/moonsniff/tbbmatch.lua [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+local free = C.rte_pktmbuf_free_export
+local band = bit.band
+
+local pktmatch = nil
+local scratchpad = nil
+local SCR_SIZE = 16 		-- size of the scratchpad in bytes, must always be multiple of 8 for hash to work
+				-- maximum: 64 (largest supported key size for hashmap)
+local mempool = nil
+local mempool2 = nil
+local next_mem = 0
```
Comment (emmericp) [inline examples/moonsniff/tbbmatch.lua]: nextMem?

File: examples/moonsniff/traffic-gen.lua [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+local PKT_LEN = 100             -- in byte
+local MS_TYPE = 0b01010101
+local band = bit.band
+
+function configure(parser)
+        parser:description("Generate traffic which can be used by moonsniff to establish latencies induced by a device under test.")
+        parser:argument("dev", "Devices to use."):args(2):convert(tonumber)
+	parser:option("-t --time", "Determines how long packets will be send in seconds."):args(1):convert(tonumber):default(10)
+	parser:option("-r --rate", "Approximate send rate in mbit/s. Due to IFG etc. rate on the wire may be higher."):args(1):convert(tonumber):default(1000)
```
Comment (emmericp) [inline examples/moonsniff/traffic-gen.lua]: ws

File: examples/moonsniff/traffic-gen.lua [hunk 6]
```diff
@@ -0,0 +1,70 @@
+--- Demonstrates and tests hardware timestamping capabilities
```
Comment (emmericp) [inline examples/moonsniff/traffic-gen.lua]: comment copy pasted from somewhere?

File: examples/moonsniff/traffic-gen.lua [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        while lm.running() and runtime:running() do
+                bufs:alloc(PKT_LEN)
+
+                for i, buf in ipairs(bufs) do
+                        local pkt = buf:getUdpPacket()
+			-- for setters to work correctly, the number is not allowed to exceed 16 bit
+			pkt.ip4:setID(band(pkt_id, 0xFFFF))
+                        pkt.payload.uint32[0] = pkt_id
+			pkt.payload.uint8[4] = MS_TYPE
+                        pkt_id = pkt_id + 1
```
Comment (emmericp) [inline examples/moonsniff/traffic-gen.lua]: ws

File: lua/moonsniff-io.lua [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+local C = ffi.C
+
+ffi.cdef[[
+	//----------------Moonsniff Live IO------------------------
+        struct ms_stats {
+                int64_t average_latency;
+		int64_t variance_latency;
+                uint32_t hits;
+                uint32_t misses;
+                uint32_t inval_ts;
```
Comment (emmericp) [inline lua/moonsniff-io.lua]: ws

File: src/hashmap.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+#include <c_bindings.h>
+#include <deque>
+
+namespace hash_map {
+    /* Secret hash cookie */
+    constexpr uint32_t secret = 0xF00BA;
+    constexpr uint64_t sip_secret[2] = {1, 2}; // 128 bit secret
+
+    template<typename K, typename std::enable_if<std::is_pod<K>::value>::type * = nullptr>
+    struct var_crc_hash {
```
Comment (emmericp) [inline src/hashmap.cpp]: maybe rename to SipHash?

File: src/moonsniff.cpp [hunk 10]
```diff
@@ -0,0 +1,122 @@
+#include <cstdint>
+#include <string>
+#include <iostream>
+#include <mutex>
+
+#define UINT24_MAX 16777215
+#define INDEX_MASK (uint32_t) 0x00FFFFFF
+#define NEGATIVE_THRESH -100 // Latencies smaller than this threshold are ignored
```
Comment (emmericp) [inline src/moonsniff.cpp]: the lua code defines this as 50ms?

File: src/moonsniff.cpp [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * Other modes are implemented in examples/moonsniff/
+ */
+namespace moonsniff {
+
+	// vars for live average computation
+	uint64_t count = 0;
+        double m2 = 0;
+        double mean = 0;
+        double variance = 0;
```
Comment (emmericp) [inline src/moonsniff.cpp]: ws

File: examples/moonsniff/arrmatch.lua [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+local TIME_THRESH = -50 	-- negative timevalues smaller than this value are not allowed
+
+-- pointers and ctypes
+local INT64_T = ffi.typeof("int64_t")
+local UINT64_T = ffi.typeof("uint64_t")
+local UINT64_P = ffi.typeof("uint64_t *")
+
+ffi.cdef[[
+	void* malloc(size_t);
+	void free(void*);
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/arrmatch.lua]: ok, will add dependency to memory.lua instead

File: examples/moonsniff/arrmatch.lua [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+
+	local precap = readSingle(prereader)
+	local postcap = readSingle(postreader)
+	log:info("Pre identifier: " .. tostring(getId(precap)) .. ", Post identifier: " .. tostring(getId(postcap)))
+
+	-- debug and information values
+	local overwrites = 0
+	local misses = 0
+	local pre_count = 0
+	local post_count = 0
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/arrmatch.lua]: renamed the entities for clarification

File: examples/moonsniff/arrmatch.lua [hunk 14]
```diff
@@ -0,0 +1,289 @@
+--- Matching for mscap files
+
+local mod = {}
+
+local hist      = require "histogram"
+local log       = require "log"
+local ms	= require "moonsniff-io"
+local bit	= require "bit"
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/arrmatch.lua]: fixed

File: examples/moonsniff/post-processing.lua [hunk 15]
```diff
@@ -0,0 +1,128 @@
+--- Demonstrates the basic usage of moonsniff in order to determine device induced latencies
+--- This file is the entry point for processing .mscap and .pcap files
+--- In order for pcap files to be processed correctly they must feature a 64 bit timestamp at the end
+---
+--- Valid files can be generated by sniffer.lua with no flags (mscap) or with the flag --capture (pcap)
+
+local log       = require "log"
+local dpdk	= require "dpdk"
+local pcap	= require "pcap"
+local profile 	= require "jit.p"
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/post-processing.lua]: fixed

File: examples/moonsniff/post-processing.lua [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+local TIME_THRESH = -50 	-- negative timevalues smaller than this value are not allowed
+
+local MODE_MSCAP, MODE_PCAP = 0, 1
+local MODE = MODE_MSCAP
+
+-- skip the initialization of DPDK, as it is not needed for this script
+dpdk.skipInit()
+
+function configure(parser)
+        parser:description("Demonstrate and test hardware latency induced by a device under test.\nThe ideal test setup is to use 2 taps, one should be connected to the ingress cable, the other one to the egress one.\n\nThe type of the input is automatically detected based on the file extension. Make sure input files are either .pcap or .mscap files.\n\nTo determine which is the pre and which is the post file, files must adhere to the following naming conventions:\n\t<name>-pre.pcap and <name>-post.pcap\n\tor\n\t<name>-pre.mscap and <name>-post.mscap\n\nNote that <name> need not be the same name.")
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/post-processing.lua]: fixed

File: examples/moonsniff/post-processing.lua [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+	parser:option("-i --input", "Path to input file. Supports .mscap or .pcap files."):args(1)
+	parser:option("-s --second-input", "Path to second input file. Supports .mscap or .pcap files."):args(1):target("second")
+	parser:option("-o --output", "Name of the histogram which is generated."):args(1):default("hist")
+	parser:option("-n --nrbuckets", "Size of a bucket for the resulting histogram."):args(1):convert(tonumber):default(1)
+	parser:flag("-d --debug", "Create debug information. Instead of processing the input files normally, they are translated into human readable csv files.")
+	parser:flag("-p --profile", "Profile the application. May decrease the overall performance.")
+        return parser:parse()
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/post-processing.lua]: fixed

File: examples/moonsniff/post-processing.lua [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+function master(args)
+	if args.input then INPUT_PATH = args.input end
+
+	-- determine the input type
+	if string.match(args.input, ".*%.pcap") then
+		MODE = MODE_PCAP
+	elseif string.match(args.input, ".*%.mscap") then
+		MODE = MODE_MSCAP
+	else
+		log:err("Input with unknown file extension.\nCan only process .pcap or .mscap files.")
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/post-processing.lua]: fixed

File: examples/moonsniff/post-processing.lua (hunk 1 above)
Comment (AP-Frank) [reply examples/moonsniff/post-processing.lua]: Probably fixed?

File: examples/moonsniff/tbbmatch.lua (hunk 4 above)
Comment (AP-Frank) [reply examples/moonsniff/tbbmatch.lua]: Used to switch between mempools

Renamed mempools and next_mem

File: examples/moonsniff/post-processing.lua (hunk 2 above)
Comment (AP-Frank) [reply examples/moonsniff/post-processing.lua]: correction: libmoon.getTime() would require DPDK startup which is unnecessary in this case

using utils: getMonotonicTime() instead

File: examples/moonsniff/arrmatch.lua [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+local UINT64_P = ffi.typeof("uint64_t *")
+
+ffi.cdef[[
+	void* malloc(size_t);
+	void free(void*);
+
+	struct entry {
+		uint64_t timestamp;	// the full timestamp
+		uint64_t identifier;	// may seem like memory waste as identifier is only 32 bit
+					// but struct is 8 byte aligned by compiler anyway
```
[the remark this answers is not in the document]
Comment (AP-Frank) [reply examples/moonsniff/arrmatch.lua]: packed seems to boost performance a little

**Comment by emmericp** [comment]
merged and squashed manually

**Pull request closed without merging**